File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/hash/Djb2.java
package com.freight.hash;

/** djb2 over raw bytes. */
public final class Djb2 implements HashAlgorithm {

    @Override
    public String name() {
        return "djb2";
    }

    @Override
    public long apply(byte[] data) {
        long state = 5381L;
        for (int i = 0; i < data.length; i++) {
            state = (state * 33L + (data[i] & 0xFF)) & 0xFFFFFFFFL;
        }
        return state;
    }
}

File: freight-triage-polyglot-ledger-recovery/environment/native/src/hash/djb2.cpp
#include "freight/hashes.h"

namespace freight {

// djb2 over raw bytes.
uint64_t hash_djb2(const std::string& data) {
  uint32_t state = 5381u;
  for (size_t i = 0; i < data.size(); ++i) {
    state = state * 33u + static_cast<uint8_t>(data[i]);
  }
  return static_cast<uint64_t>(state);
}

}  // namespace freight
